plugin/pkg/admission/podtopologylabels/admission.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podtopologylabels
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/admission"
+	genericadmissioninitializer "k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/client-go/informers"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/component-base/featuregate"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+// PluginName is a string with the name of the plugin
+const PluginName = "PodTopologyLabels"
+
+// defaultConfig is the configuration used for the default instantiation of the plugin.
+// This is the configured used by kube-apiserver.
+// It is not exported to avoid any chance of accidentally mutating the variable.
+var defaultConfig = Config{
+	Labels: []string{"topology.k8s.io/zone", "topology.k8s.io/region", "kubernetes.io/hostname"},
+}
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(_ io.Reader) (admission.Interface, error) {
+		plugin := NewPodTopologyPlugin(defaultConfig)
+		return plugin, nil
+	})
+}
+
+// Config contains configuration for instances of the podtopologylabels admission plugin.
+// This is not exposed as user-facing APIServer configuration, however can be used by
+// platform operators when building custom topology label plugins.
+type Config struct {
+	// Labels is set of explicit label keys to be copied from the Node object onto
+	// pod Binding objects during admission.
+	Labels []string
+	// Domains is a set of label key prefixes used to copy across label values
+	// for all labels with a given domain prefix.
+	// For example, `example.com` would match all labels matching `example.com/*`.
+	// Keys without a domain portion (i.e. those not containing a /) will never match.
+	Domains []string
+	// Suffixes is a set of label key domain suffixes used to copy label values for
+	// all labels of a given subdomain.
+	// This acts as a suffix match on the domain portion of label keys.
+	// If a suffix does not have a leading '.', one will be prepended to avoid
+	// programmer errors with values like `example.com` matching `notexample.com`.
+	// Keys without a domain portion (i.e. those not containing a /) will never match.
+	Suffixes []string
+}
```

Comment:
I purposely left the `Domains` and `Suffixes` support in here in case an external plugin wants to make use of this kind of 'matching', but it isn't actually used in-tree (see `defaultConfig` above, which only sets `Labels`) and could be seen as unnecessary bloat/code to maintain.

Any particularly strong opinions on this? I can remove it if preferable :)

Reply:
Maybe we introduce it here later when there is a need.

Reply:
removed this code - can be added if and when required :)